### PR TITLE
Support non-ASCII characters in function arguments

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Support non-ASCII characters in PyGMT arguments, so that users don't have to use octal codes. 
See https://github.com/GenericMappingTools/pygmt/issues/2204 for context.

**Working example**

```python
import pygmt

fig = pygmt.Figure()
fig.basemap(region=[0, 10, 0, 5], projection="x1c", frame="WSEN+tABC¥±°″≥×≈∇⇔∑DEF")
fig.show()
```
it produces:

map

**Notes to maintainers**
The symbols and ZapfDingbats charset can be obtained from https://unicode.org/Public/MAPPINGS/VENDORS/ADOBE/zdingbat.txt and https://unicode.org/Public/MAPPINGS/VENDORS/ADOBE/symbol.txt. The following script is used to generate the list of characters, but manually editing are necessary.
```
with open("zdingbat.txt", "r") as fin:
    for line in fin:
        if line.startswith("#"):
            continue
        hexval = line.split()[0].strip()
        print(chr(int(hexval, base=16)), end="")
        i += 1
        if i % 16 == 0:
            print()
```

**TODO**

- [x] Support all non-ASCII characters in the ISOLatin1+ character set
- [x] Support non-ASCII characters in the Symbols character set
- [x] Support non-ASCII characters in the ZapfDingbats character set
- [ ] Check `PS_CHAR_ENCODING` so it also works with Standard+ character set?
- [ ] Check arguments that are not processed by `build_arg_string` (e.g., `text` in `Figure.text`)
- [ ] Add a gallery example or a tutorial?
- [x] update any existing examples

Closes https://github.com/GenericMappingTools/pygmt/issues/2204 Closes https://github.com/GenericMappingTools/pygmt/issues/2000

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/helpers/utils.py [hunk 1]
```diff
@@ -91,6 +91,41 @@ def data_kind(data, x=None, y=None, z=None, required_z=False):
     return kind
 
 
+def non_ascii_to_octal(argstr):
+    r"""
+    Translate non-ASCII characters to their corresponding octal codes.
+
+    Currently, only ISOLatin1+ character set is supported.
```
Comment (yvonnefroehlich) [inline pygmt/helpers/utils.py:98]:
```suggestion
    Currently, only the ISOLatin1+ character set is supported.
```

**Comment by seisman** [comment]
This new feature is almost done and is ready for more testings. Please let me know what you think.

**Comment by weiji14** [comment]
Looking good! Could you also change this line in the colorbar gallery example to use degrees perhaps?

https://github.com/GenericMappingTools/pygmt/blob/dde23f11f309ecafa51262149ecb13c083812728/examples/gallery/embellishments/colorbar.py#L43

Also, do we need to test different [PS_CHAR_ENCODING](https://docs.generic-mapping-tools.org/6.4/gmt.conf.html#term-PS_CHAR_ENCODING) settings like Standard, Standard+, ISOLatin1, ISOLatin1+, and ISO-8859-x?

**Comment by seisman** [comment]
> Looking good! Could you also change this line in the colorbar gallery example to use degrees perhaps?
> 
> https://github.com/GenericMappingTools/pygmt/blob/dde23f11f309ecafa51262149ecb13c083812728/examples/gallery/embellishments/colorbar.py#L43
> 

Done in e1b43b2eadd6923064d8faf5b1b4ad7b58929c81.

> Also, do we need to test different [PS_CHAR_ENCODING](https://docs.generic-mapping-tools.org/6.4/gmt.conf.html#term-PS_CHAR_ENCODING) settings like Standard, Standard+, ISOLatin1, ISOLatin1+, and ISO-8859-x? Currently the mapping seems to assume ISOLatin1, but if a user sets PS_CHAR_ENCODING to another encoding, the output might be different.

ISOLatin1+ is an extension of ISOLatin1, which means ISOLatin1+ supports more characters than ISOLatin (https://docs.generic-mapping-tools.org/latest/cookbook/octal-codes.html). Thus, it does no harm to support ISOLatin1+ only characters even if users set ISOLatin1 encoding.

For Standard+ and other ISO-8859-x encodings, yes, we can definitely support them. But it also means we have to carefully check their differences and maintain the corresponding mapping dictionary from character to octal codes.

File: pygmt/helpers/utils.py [hunk 2]
```diff
@@ -91,6 +92,84 @@ def data_kind(data, x=None, y=None, z=None, required_z=False):
...
+        {
+            c: "\\" + format(i, "o")
+            for c, i in zip(
+                "•…™—–ﬁžšŒ†‡Ł⁄‹Š›œŸŽł‰„“”ı`´ˆ˜¯˘˙¨‚˚¸'˝˛ˇ",
+                [*range(25, 32), *range(127, 160)],
+            )
+        }
+    )
+    # ISOLatin1+ charset: \240-\377
+    mapping.update({chr(i): "\\" + format(i, "o") for i in range(160, 256)})
```
Comment (seisman) [inline pygmt/helpers/utils.py:166]: Actually, Python supports many different encodings, thus it's possible to implement this feature without manually maintaining the big dictionary, just like what I already do at line 166 for ISOLatin1+ characters \240 to \377.

File: pygmt/helpers/utils.py [hunk 3]
```diff
@@ -91,6 +92,84 @@ def data_kind(data, x=None, y=None, z=None, required_z=False):
...
+    -------
+    str
+        The translated string.
+
+    Examples
+    --------
+    >>> non_ascii_to_octal("•‰“”±°ÿ")
+    '\\31\\214\\216\\217\\261\\260\\377'
+    >>> non_ascii_to_octal("αζ∆Ω∑∏∇")
+    '@~\\141@~@~\\172@~@~\\104@~Ω@~\\345@~@~\\325@~@~\\321@~
```
Comment (seisman) [inline pygmt/helpers/utils.py:122]:
```suggestion
    '@~\\141@~@~\\172@~@~\\104@~Ω@~\\345@~@~\\325@~@~\\321@~'
```

File: pygmt/helpers/utils.py (hunk 2 above)
Comment (seisman) [reply pygmt/helpers/utils.py]: **Update:**

Can be done using:
```python
for i in range(160, 256):
    print(i, chr(i), chr(i).encode("iso-8859-1").decode("iso-8859-5"))
```

**Comment by seisman** [comment]
> For Standard+ and other ISO-8859-x encodings, yes, we can definitely support them. But it also means we have to carefully check their differences and maintain the corresponding mapping dictionary from character to octal codes.

For other ISO-8859-x encodings, it's possible to generate the big mapping dictionary using the following codes:

```
for i in range(160, 256):
    print(i, chr(i), chr(i).encode("iso-8859-1").decode("iso-8859-5"))
```

File: examples/gallery/embellishments/colorbar.py [hunk 4]
```diff
@@ -40,7 +40,7 @@
     # with a length/width (+w) of 4 cm by 0.5 cm, and plotted horizontally (+h)
     position="g0.3/8.7+w4c/0.5c+h",
     box=True,
-    frame=["x+lTemperature", r"y+l\260C"],
+    frame=["x+lTemperature", r"y+l°C"],
```
Comment (yvonnefroehlich) [inline examples/gallery/embellishments/colorbar.py:43]: I feel the leading `r` is not anymore needed 🙂, or?
```suggestion
    frame=["x+lTemperature", "y+l°C"],
```

File: pygmt/helpers/utils.py [hunk 5]
```diff
@@ -145,6 +146,119 @@ def data_kind(data=None, x=None, y=None, z=None, required_z=False):
...
+                + "≅ΑΒΧΔΕΦΓΗΙϑΚΛΜΝΟ"  # \10x-11x
+                + "ΠΘΡΣΤΥςΩΞΨΖ[∴]⊥_"  # \12x-13x
+                + "αβχδεφγηιϕκλμνο"  # \14x-15x
+                + "πθρστυϖωξψζ{|}∼"  # \16x-17x. \177 is undefined
+                + "€ϒ′≤⁄∞ƒ♣♦♥♠↔←↑→↓"  # \24x-\25x
+                + "°±″≥×∝∂•÷≠≡≈…↵"  # \26x-27x
+                + "ℵℑℜ℘⊗⊕∅∩∪⊃⊇⊄⊂⊆∈∉"  # \30x-31x
+                + "∠∇®©™∏√⋅¬∧∨⇔⇐⇑⇒⇓"  # \32x-33x
+                + "◊〈®©™∑"  # \34x-35x
+                + "〉∫⌠⌡",  # \36x-37x. \360 and \377 are undefined
```
Comment (seisman) [inline pygmt/helpers/utils.py:206]: Symbols \140, \275 and \276 don't exist in the unicode table, so they're incorrectly shown.

Some symbols of \36x and \37x are shown as boxes in the GitHub web GUI, but they're shown correctly in Vim, so these symbols should be good.

Screenshot from 2023-08-02 15-08-53

**Comment by seisman** [comment]
> Check PS_CHAR_ENCODING so it also works with Standard+ character set?

Already answered in https://github.com/GenericMappingTools/pygmt/pull/2584#issuecomment-1616411916.

> For other ISO-8859-x encodings, it's possible to generate the big mapping dictionary automatically using the following code:
> ```
> {chr(i).encode("iso-8859-1").decode("iso-8859-5"): "\\" + format(i, "o") for i in range(160, 256)}
> ```
> but we have to check the PS_CHAR_ENCODING parameter to decide which dictionary to use. However, adding `from pygmt.clib.session import Session` to `pygmt.helpers.utils` results in circular imports.
> 

> Check arguments that are not processed by build_arg_string (e.g., `text` in Figure.text)

Yes, we can also translate the strings passed by the `text` parameter, but we can do very little to plaintext files (we can, but it means we have to open and read the whole plaintext file, which is not efficient). So not sure what we should do here.

> Add a gallery example or a tutorial?

Better to do it in a separate PR.

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers Any comments and suggestions?

**Review by michaelgrund — APPROVED** [review]
Looks good to me!

File: pygmt/helpers/utils.py [hunk 6]
```diff
@@ -196,6 +197,119 @@ def data_kind(data=None, x=None, y=None, z=None, required_z=False, required_data
...
+
+    Parameters
+    ----------
+    argstr : str
+        The string to be translated.
+
+    Returns
+    -------
+    str
+        The translated string.
```
Comment (weiji14) [inline pygmt/helpers/utils.py:215]:
```suggestion
    translated_argstr : str
        The translated string.
```

**Review by weiji14 — APPROVED** [review]
Haven't checked every individual symbol, but I trust that they are correct :slightly_smiling_face:

> > Check arguments that are not processed by build_arg_string (e.g., `text` in Figure.text)
> 
> Yes, we can also translate the strings passed by the `text` parameter, but we can do very little to plaintext files (we can, but it means we have to open and read the whole plaintext file, which is not efficient). So not sure what we should do here.

Ok, to handle ASCII characters in `fig.text` in a separate PR.

File: pygmt/helpers/utils.py (hunk 2 above)
Comment (seisman) [reply pygmt/helpers/utils.py]: Here is the improved version:
```
for code in [*range(0o040, 0o200), *range(0o240, 0o400)]:
    char = codecs.decode(bytes([code]), f"iso8859-{i}", errors="replace")
```

**Pull request merged**